lib/Sema/CSSimplify.cpp

```diff
@@ -1212,10 +1212,8 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
     return result;
 
   // Result type can be covariant (or equal).
-  return matchTypes(func1->getResult(), func2->getResult(), subKind,
-                     subflags,
-                     locator.withPathElement(
-                       ConstraintLocator::FunctionResult));
+  return matchTypes(func1->getResult(), func2->getResult(), subKind, subflags,
```

Comment:
Is there any way to decide here we have a ClosureResult, avoiding the below check of the type variable’s locator?

Reply:
Maybe I should revisit that since I've changed ranking, originally I've tried to use `ClosureResult` instead of `FunctionResult` locator while creating this `matchTypes` instead why `clang-format` changed formatting here but it had side-effects in other parts of the solver.

Reply:
So yeah, it looks like it still changes behavior:

```
class SomeClass {}
class SomeSpecificClass : SomeClass {}
func takesSomeClassGenerator(_ fn : () -> SomeClass) {}

func generateWithConstant(_ x : SomeSpecificClass) {
  takesSomeClassGenerator({ x })
}
```

closure is going to infer as `() -> SomeSpecificClass` which would result in thunk from `() -> SomeSpecificClass` to `() -> SomeClass`.

Reply:
This is a similar problem to

```
func id<T>(_ t: T) -> T { return t }

func generateWithConstant(_ x : SomeSpecificClass) -> SomeClass {
  return id(x)
}
```

What do we infer as `T` there? We can say it doesn't matter, or we can come up with some consistent rule, but it might be hard.

Reply:
`T` inferred as `SomeSpecificClass` in this case _currently_.

Reply:
Right, but you could easily argue it should be the other way too :)

Reply:
I know :) but what's concerning is that test I've pasted originally was added specifically to test that we end up without conversion, is it because thunks are expensive?

Reply:
Ideally we would not emit thunks that wrap literal closures, but it's not the end of the world if we do.